neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -75,35 +87,139 @@ def image_input(self) -> tf.Tensor:
             name="input_images")
 
     @tensor
-    def image_processing_layers(self) -> List[tf.Tensor]:
+    def image_mask(self) -> tf.Tensor:
+        return tf.placeholder(
+            tf.float32,
+            shape=(None, self.image_height, self.image_width, 1),
+            name="input_mask")
+
+    def plain_convolution(
+            self,
+            prev_layer: tf.Tensor,
+            prev_mask: tf.Tensor,
+            specification: ConvSpec,
+            layer_num: int) -> Tuple[tf.Tensor, tf.Tensor, int]:
+        if len(specification) != 5:
+            raise ValueError((
+                "Specification of a convolutional layer needs to "
+                "have 5 members: kernel size, stride, padding, "
+                "# output channels, was {}").format(specification))
+        kernel_size, stride, pad, out_channels = specification[1:]
+
+        if pad not in ["same", "valid"]:
```

Comment:
case insensitive?

Reply:
A potřebujeme tu kontrolu vůbec? Za tři další řádky se volá funkce z TF a ta hodí úplně stejnou hlášku.

Reply:
Já bych já tu nechal, protože v tom TF se ještě zavolá bůhví kolik funkcí.

Reply:
tf layers conv2d volá konstruktor Conv2D, kterej jako první kománd volá konstruktor předka, kterej jako první věc normalizuje parametry a normalizace paddingu ti hodí přesně tuhle chybu.

Reply:
taky se to tam lowercasuje

Reply:
(lžu, Conv_ ještě volá konstruktor svýho předka, kterej toho nicméně moc nedělá)

Reply:
ale klidně to tu nech